[WIP][inductor] better heuristics found by coordesc

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97855
* #98524

I'll accumulate the perf numbers here: https://docs.google.com/spreadsheets/d/1Ga5_twazRahmS5M0g8CzCapZfB4cepy-265ffYrIjPE/edit#gid=0 



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @soumith @anijain2305 @desertfire

shunting314 committed:
[inductor] better heuristics found by coordesc

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] better heuristics found by coordesc"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] better heuristics found by coordesc"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] better heuristics found by coordesc"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] better heuristics found by coordesc"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] better heuristics found by coordesc"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] better heuristics found by coordesc"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] better heuristics found by coordesc"


I'll accumulate the perf numbers here: https://docs.google.com/spreadsheets/d/1Ga5_twazRahmS5M0g8CzCapZfB4cepy-265ffYrIjPE/edit#gid=0 



cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] better heuristics found by coordesc"


I'll accumulate the perf numbers here: https://docs.google.com/spreadsheets/d/1Ga5_twazRahmS5M0g8CzCapZfB4cepy-265ffYrIjPE/edit#gid=0 



cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ngimel commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -514,7 +509,17 @@ def pointwise(size_hints, meta, tile_hint=None, filename=None):
+        configs = [triton_config(size_hints, bs)]
+        if config.max_autotune:
+            configs.extend([
+                # improve 1.832x for https://gist.github.com/shunting314/69b5055193148ade349ac7e58c85d2d9 
```
this case is probably spilling a lot with default config, we should check spills when we compile and recompile with more warps, even when max_autotune is not on.

shunting314 replied:
yep, I also checked the kernel's triton code. It does a lot of things. The default config simply based on numel can pick a too large block size.

I'll followup with a separate PR to consider number of spills

ngimel commented on torch/_inductor/utils.py:
```diff
@@ -912,7 +912,10 @@ def report_category(category, profile_events):
             percent = f"{ev.self_cuda_time_ms / wall_time_ms * 100:.2f}%"
-            rows.append([ev.key[:120], ev.self_cuda_time_ms, ev.count, percent])
+            key = ev.key
+            if key.startswith("triton_poi_fused_"):
```
why this change?

shunting314 replied:
Some kernel name is too long to fit the screen. Remove the common/uninteresting prefix.

ngimel commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -582,6 +587,18 @@ def reduction(size_hints, reduction_hint=False, meta=None, filename=None):
+                # improve 1.143x for https://gist.github.com/shunting314/ae6a0d3c63409bed7759ee7bdcde01a8
+                Config({"XBLOCK": 1, "RBLOCK": 1024}, num_warps=16, num_stages=1),
+                # improve 1.033x for https://gist.github.com/shunting314/d1036afd5210e19da3144c24f3b475d8
+                Config({"XBLOCK": 1, "RBLOCK": 1024}, num_warps=8, num_stages=1),
```
here again we can probably rely on spills information to not check configs with insufficient num_warps?

ngimel reviewed (approved):
I'm not against landing as is, however we should come up with ways to filter configs after compilation (based on spills) without full benchmarking

shunting314 commented:
> I'm not against landing as is, however we should come up with ways to filter configs after compilation (based on spills) without full benchmarking

So far in this PR I just add the promising configs found by coordinate descent tuning when applied to the top expensive pointwise and reduction kernels of a model/graph. I'll add more as I tried on more models.

The purpose is to see what overall perf improvement we can get by adding more configs (disregarding compiling time at the beginning). This is important since we don't want to improve individual kernel perf a lot but dost not improve the model training time overall 

I'll try to use other information (like number of spills you mentioned) to reduce the number of configs we should benchmark.

ngimel commented:
You can get perf on this PR using ciflow/inductor-perf-compare label, but I don't think it sets max-autotuning

shunting314 commented:
> You can get perf on this PR using ciflow/inductor-perf-compare label, but I don't think it sets max-autotuning

Thanks for mentioning this. I didn't know we have such a label.

Does it run with and without the PR and compare the perf?

shunting314 committed:
Update on "[WIP][inductor] better heuristics found by coordesc"


I'll accumulate the perf numbers here: https://docs.google.com/spreadsheets/d/1Ga5_twazRahmS5M0g8CzCapZfB4cepy-265ffYrIjPE/edit#gid=0 



cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] better heuristics found by coordesc"


I'll accumulate the perf numbers here: https://docs.google.com/spreadsheets/d/1Ga5_twazRahmS5M0g8CzCapZfB4cepy-265ffYrIjPE/edit#gid=0 



cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] better heuristics found by coordesc"


I'll accumulate the perf numbers here: https://docs.google.com/spreadsheets/d/1Ga5_twazRahmS5M0g8CzCapZfB4cepy-265ffYrIjPE/edit#gid=0 



cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] better heuristics found by coordesc"


I'll accumulate the perf numbers here: https://docs.google.com/spreadsheets/d/1Ga5_twazRahmS5M0g8CzCapZfB4cepy-265ffYrIjPE/edit#gid=0 



cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] better heuristics found by coordesc"


I'll accumulate the perf numbers here: https://docs.google.com/spreadsheets/d/1Ga5_twazRahmS5M0g8CzCapZfB4cepy-265ffYrIjPE/edit#gid=0 



cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] better heuristics found by coordesc"


I'll accumulate the perf numbers here: https://docs.google.com/spreadsheets/d/1Ga5_twazRahmS5M0g8CzCapZfB4cepy-265ffYrIjPE/edit#gid=0 



cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] better heuristics found by coordesc"


I'll accumulate the perf numbers here: https://docs.google.com/spreadsheets/d/1Ga5_twazRahmS5M0g8CzCapZfB4cepy-265ffYrIjPE/edit#gid=0 



cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] better heuristics found by coordesc"


I'll accumulate the perf numbers here: https://docs.google.com/spreadsheets/d/1Ga5_twazRahmS5M0g8CzCapZfB4cepy-265ffYrIjPE/edit#gid=0 



cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] better heuristics found by coordesc"


I'll accumulate the perf numbers here: https://docs.google.com/spreadsheets/d/1Ga5_twazRahmS5M0g8CzCapZfB4cepy-265ffYrIjPE/edit#gid=0 



cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] better heuristics found by coordesc"


I'll accumulate the perf numbers here: https://docs.google.com/spreadsheets/d/1Ga5_twazRahmS5M0g8CzCapZfB4cepy-265ffYrIjPE/edit#gid=0 



cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] better heuristics found by coordesc"


I'll accumulate the perf numbers here: https://docs.google.com/spreadsheets/d/1Ga5_twazRahmS5M0g8CzCapZfB4cepy-265ffYrIjPE/edit#gid=0 



cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] better heuristics found by coordesc"


I'll accumulate the perf numbers here: https://docs.google.com/spreadsheets/d/1Ga5_twazRahmS5M0g8CzCapZfB4cepy-265ffYrIjPE/edit#gid=0 



cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] better heuristics found by coordesc"


I'll accumulate the perf numbers here: https://docs.google.com/spreadsheets/d/1Ga5_twazRahmS5M0g8CzCapZfB4cepy-265ffYrIjPE/edit#gid=0 



cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]
